File: baml_language/sdk_tests/crates/cpp/type_shapes/customizable/tests/test_literals.cc
// Roundtrip coverage for baml_sdk::literals - literal Ty variants as
// singleton ::baml::lit types. Port of roundtrip_tests/test_literals.py:
// python keeps the value set in typing.Literal annotations; C++ makes each
// value a distinct type, so the round trips construct through the
// BAML_LIT macro family and read back through Lit's implicit value
// conversions.
#include <baml_sdk.h>
#include <baml_test.h>

#include <string>
#include <string_view>
#include <type_traits>
#include <variant>

using baml_sdk::literals::Literals;

// The generated surface is typed per value, not widened.
static_assert(std::is_same<decltype(baml_sdk::literals::return_literal42()),
                           BAML_LIT(42)>::value,
              "int literal return must be a Lit");
static_assert(std::is_same<decltype(baml_sdk::literals::return_literal_draft()),
                           BAML_LIT("draft")>::value,
              "string literal return must be a Lit");
static_assert(std::is_same<decltype(baml_sdk::literals::return_literal_true()),
                           BAML_LIT(true)>::value,
              "bool literal return must be a Lit");

BAML_TEST(literals_return_literals) {
  BAML_ASSERT(baml_sdk::literals::return_literal42() == BAML_LIT(42){});
  BAML_ASSERT(baml_sdk::literals::return_literal_neg_one() == BAML_LIT(-1){});
  BAML_ASSERT(baml_sdk::literals::return_literal_draft() ==
              BAML_LIT("draft"){});
  BAML_ASSERT(baml_sdk::literals::return_literal_escaped() ==
              BAML_LIT("has \"quotes\""){});
  BAML_ASSERT(baml_sdk::literals::return_literal_true() == BAML_LIT(true){});
  BAML_ASSERT(baml_sdk::literals::return_literal_false() == BAML_LIT(false){});
}

BAML_TEST(literals_literal_values_convert_implicitly) {
  const int64_t n = baml_sdk::literals::return_literal42();
  BAML_ASSERT(n == 42);
  const std::string_view s = baml_sdk::literals::return_literal_draft();
  BAML_ASSERT(s == "draft");
  const bool b = baml_sdk::literals::return_literal_true();
  BAML_ASSERT(b);
  BAML_ASSERT(BAML_LIT("has \"quotes\"")::value == "has \"quotes\"");
}

BAML_TEST(literals_round_trip_literal42) {
  BAML_ASSERT(baml_sdk::literals::round_trip_literal42(BAML_LIT(42){}) ==
              BAML_LIT(42){});
}

BAML_TEST(literals_round_trip_literal_draft) {
  BAML_ASSERT(baml_sdk::literals::round_trip_literal_draft(
                  BAML_LIT("draft"){}) == BAML_LIT("draft"){});
}

BAML_TEST(literals_round_trip_literal_escaped) {
  BAML_ASSERT(baml_sdk::literals::round_trip_literal_escaped(BAML_LIT(
                  "has \"quotes\""){}) == BAML_LIT("has \"quotes\""){});
}

BAML_TEST(literals_round_trip_literal_true) {
  BAML_ASSERT(baml_sdk::literals::round_trip_literal_true(BAML_LIT(true){}) ==
              BAML_LIT(true){});
}

BAML_TEST(literals_round_trip_literal_false) {
  BAML_ASSERT(baml_sdk::literals::round_trip_literal_false(BAML_LIT(false){}) ==
              BAML_LIT(false){});
}

BAML_TEST(literals_round_trip_literals) {
  const Literals lit{BAML_LIT(42){}, BAML_LIT("draft"){},
                     BAML_LIT("has \"quotes\""){}, BAML_LIT(true){},
                     BAML_LIT(false){}};
  BAML_ASSERT(baml_sdk::literals::round_trip_literals(lit) == lit);
}

BAML_TEST(literals_round_trip_flag_mixed_literal_union) {
  // Mixed-base literal union ("active" | 1 | true): the union codec's
  // literal pass must route each wire arm to its exact Lit alternative.
  using baml_sdk::literals::Flag;
  Flag f = baml_sdk::literals::round_trip_flag(BAML_LIT("active"){});
  BAML_ASSERT(std::holds_alternative<BAML_LIT("active")>(f));
  f = baml_sdk::literals::round_trip_flag(BAML_LIT(1){});
  BAML_ASSERT(std::holds_alternative<BAML_LIT(1)>(f));
  f = baml_sdk::literals::round_trip_flag(BAML_LIT(true){});
  const int which = baml::match(
      f,                                                                    //
      [](BAML_LIT("active")) { return 0; }, [](BAML_LIT(1)) { return 1; },  //
      [](BAML_LIT(true)) { return 2; });
  BAML_ASSERT(which == 2);
}

File: baml_language/sdk_tests/crates/python_pydantic2/type_shapes/customizable/roundtrip_tests/test_literals.py
"""Roundtrip coverage for `baml_sdk.literals` — literal Ty variants.

Float literals are intentionally absent (Python `Literal` rejects
floats). The negative-literal-as-field case is still parser-blocked, but
the function-return form `return_literal_neg_one() -> -1` emits and is
exercised here.
"""

import baml_sdk  # noqa: F401  — initializes the BAML runtime
from baml_sdk.literals import (
    Literals,
    return_literal42,
    return_literal_neg_one,
    return_literal_draft,
    return_literal_escaped,
    return_literal_true,
    return_literal_false,
    round_trip_literal42,
    round_trip_literal_draft,
    round_trip_literal_escaped,
    round_trip_literal_true,
    round_trip_literal_false,
    round_trip_literals,
)


def test_literals_return_literals():
    assert return_literal42() == 42
    assert return_literal_neg_one() == -1
    assert return_literal_draft() == "draft"
    assert return_literal_escaped() == 'has "quotes"'
    assert return_literal_true() is True
    assert return_literal_false() is False


def test_literals_round_trip_literal42():
    assert round_trip_literal42(x=42) == 42


def test_literals_round_trip_literal_draft():
    assert round_trip_literal_draft(x="draft") == "draft"


def test_literals_round_trip_literal_escaped():
    assert round_trip_literal_escaped(x='has "quotes"') == 'has "quotes"'


def test_literals_round_trip_literal_true():
    assert round_trip_literal_true(x=True) is True


def test_literals_round_trip_literal_false():
    assert round_trip_literal_false(x=False) is False


def test_literals_round_trip_literals():
    lit = Literals(
        literal_42=42,
        literal_draft="draft",
        literal_escaped='has "quotes"',
        literal_true=True,
        literal_false=False,
    )
    assert round_trip_literals(l=lit) == lit


def test_literals_round_trip_flag_mixed_literal_union():
    from baml_sdk.literals import round_trip_flag

    assert round_trip_flag(f="active") == "active"
    r_int = round_trip_flag(f=1)
    assert r_int == 1 and not isinstance(r_int, bool)
    r_bool = round_trip_flag(f=True)
    assert r_bool is True
